### PR TITLE
search  installed MS deploy first

### PR DESCRIPTION
**Pull request by Ajay-MS**

### CODE REVIEW

File: Tasks/AzureRmWebAppDeployment/Utility.ps1 [hunk 1]
```diff
@@ -2,10 +2,26 @@ $ErrorActionPreference = 'Stop'
...
+    {
+         Write-Verbose ("MSDeploy is not installed in system." + $_.Exception.Message)
+    }
+
+    if( $msDeployExePath -eq $null )
+    {
 
-    $msDeployExeDir = Join-Path $currentDir "MSDeploy3.6"
-    $msDeployExePath = Join-Path $msDeployExeDir "msdeploy.exe"
+        Write-Verbose  (Get-LocalizedString -Key "Using local MSDeploy.exe")
```
Comment (KrishnaAdityaB) [inline Tasks/AzureRmWebAppDeployment/Utility.ps1]: where is this added to be localized?

File: Tasks/AzureRmWebAppDeployment/Utility.ps1 [hunk 2]
```diff
@@ -2,10 +2,26 @@ $ErrorActionPreference = 'Stop'
...
+    }
+
+    if( $msDeployExePath -eq $null )
+    {
 
-    $msDeployExeDir = Join-Path $currentDir "MSDeploy3.6"
-    $msDeployExePath = Join-Path $msDeployExeDir "msdeploy.exe"
+        Write-Verbose  (Get-LocalizedString -Key "Using local MSDeploy.exe")  
+        $currentDir = (Get-Item -Path ".\").FullName
+        $msDeployExeDir = Join-Path $currentDir "MSDeploy3.6"
```
Comment (KrishnaAdityaB) [inline Tasks/AzureRmWebAppDeployment/Utility.ps1]: doesn't it vary based on the version?
If we support only this version - pass this as a parameter to the Get-MSDeployOnTargetMachine cmdlet

File: Tasks/AzureRmWebAppDeployment/FindInstalledMSDeploy.ps1 [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    }
+    catch [System.Exception]
+    {
+        Write-Verbose ("Failed to get MS Deploy location with exception: " + $_.Exception.Message)
+        $MSDeployVersionNumber = 0
+    }
+
+    if ($MSDeployPath -eq $null)
+    {
+        throw  "Unable to find the location of MS Deploy from registry on machine $env:COMPUTERNAME"
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/FindInstalledMSDeploy.ps1]: Should it be throw or verbose message, as we want to continue execution with pre bundled msdeploy

File: Tasks/AzureRmWebAppDeployment/Utility.ps1 [hunk 4]
```diff
@@ -2,10 +2,26 @@ $ErrorActionPreference = 'Stop'
...
-    $currentDir = (Get-Item -Path ".\").FullName
+    $msDeployExePath = $null
+    try
+    {
+        $msDeployExePath = Get-MSDeployOnTargetMachine
+    }
+    catch [System.Exception]
+    {
+         Write-Verbose ("MSDeploy is not installed in system." + $_.Exception.Message)
+    }
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/Utility.ps1]: Can you take a look at the Get-MsDeployLocation in IIS Task msdeployonTargetMachines.ps1 file. You can do it in optimized way.

File: Tasks/AzureRmWebAppDeployment/Utility.ps1 [hunk 5]
```diff
@@ -2,13 +2,33 @@ $ErrorActionPreference = 'Stop'
...
 
-    $msDeployExeDir = Join-Path $currentDir "MSDeploy3.6"
-    $msDeployExePath = Join-Path $msDeployExeDir "msdeploy.exe"
-    Write-Host (Get-LocalizedString -Key "msdeploy.exe is located at '{0}'" -ArgumentList $msDeployExePath)
+    if( $MSDeployExePath -ne $null -and $MSDeployVersion -lt 3 ){
+        throw  "Unsupported installed version found for MSDeploy"
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/Utility.ps1]: use get-localized to display

File: Tasks/AzureRmWebAppDeployment/Utility.ps1 [hunk 6]
```diff
@@ -2,13 +2,33 @@ $ErrorActionPreference = 'Stop'
...
+    }
+    catch [System.Exception]
+    {
+         Write-Verbose ("MSDeploy is not installed in system." + $_.Exception.Message)
+    }
 
-    $msDeployExeDir = Join-Path $currentDir "MSDeploy3.6"
-    $msDeployExePath = Join-Path $msDeployExeDir "msdeploy.exe"
-    Write-Host (Get-LocalizedString -Key "msdeploy.exe is located at '{0}'" -ArgumentList $msDeployExePath)
+    if( $MSDeployExePath -ne $null -and $MSDeployVersion -lt 3 ){
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/Utility.ps1]: use string::isnullorempty() for null checks

File: Tasks/AzureRmWebAppDeployment/FindInstalledMSDeploy.ps1 [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+            }
+            catch {}
+        }
+   
+    $targetKeys    
+}
+
+function Get-MSDeployPackage([int] $majorVersion, [bool] $wow6432Node)
+{
+    $MSDeployInstallRootRegKey = "SOFTWARE", "Microsoft", "IIS Extensions", "MSDeploy", "$majorVersion" -join [System.IO.Path]::DirectorySeparatorChar
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/FindInstalledMSDeploy.ps1]: make sure verbose logs are enough ti check whether we are looking things for 32-bit or 64-bit.. applicable everywher

File: Tasks/AzureRmWebAppDeployment/FindInstalledMSDeploy.ps1 [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    $keys = Get-Item $MSDeployRegKey | %{$_.GetSubKeyNames()} 
+    $versions = Get-SubKeysInFloatFormat $keys | Sort-Object -Descending
+
+    Write-Verbose "MSDeploy Versions installed on machine $env:COMPUTERNAME as read from registry: $versions"        
+
+    foreach ($majorVersion in $versions) 
+    {
+        $MSDeployInstallPathWow6432Node = Get-MSDeployPackage $majorVersion $true
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/FindInstalledMSDeploy.ps1]: we are resolving  conflicting dlls issue by picking latest version of msdeploy installed in machine. please verify whether when we installed msdeploy, dlls are placed in GAC compulsory or for different version they are overridded if it is already present in GAC

**Pull request merged**